File: apps/pharma/src/lib/services/agents/patterns/constitutional-ai.ts
/**
 * Constitutional AI Pattern Implementation
 * 
 * Implements self-critique and improvement based on constitutional principles.
 * Healthcare-specific constitution ensures medical accuracy, safety, and compliance.
 * 
 * Features:
 * - Principle-based self-review
 * - Violation detection and explanation
 * - Automatic response revision
 * - Compliance scoring
 * - Healthcare-specific principles (HIPAA, accuracy, safety)
 * 
 * Architecture Principles:
 * - SOLID: Single responsibility, dependency injection, clean interfaces
 * - Type Safety: Full TypeScript with discriminated unions
 * - Observability: Structured logging, distributed tracing
 * - Resilience: Circuit breakers, retries, graceful degradation
 * - Security: Healthcare compliance principles enforced
 * 
 * @module lib/services/agents/patterns/constitutional-ai
 * @version 1.0.0
 */

import { ChatOpenAI, ChatAnthropic } from '@langchain/openai';
import { BaseMessage, HumanMessage, SystemMessage } from '@langchain/core/messages';
import { createLogger } from '@/lib/services/observability/structured-logger';
import { getTracingService } from '@/lib/services/observability/tracing';
import { withRetry } from '@/lib/services/resilience/retry';
import { getSupabaseCircuitBreaker } from '@/lib/services/resilience/circuit-breaker';

// ============================================================================
// TYPE DEFINITIONS
// ============================================================================

/**
 * Constitutional principle
 */
export interface Principle {
  id: string;
  description: string;
  weight: number; // 1-5, higher = more important
  category: 'accuracy' | 'safety' | 'privacy' | 'ethics' | 'compliance';
}

/**
 * Principle violation
 */
export interface Violation {
  principleId: string;
  principleDescription: string;
  severity: 'low' | 'medium' | 'high' | 'critical';
  explanation: string;
  weight: number;
}

/**
 * Constitutional review result
 */
export interface ConstitutionalReview {
  violations: Violation[];
  complianceScore: number; // 0-1, higher = more compliant
  overallCompliance: 'compliant' | 'minor_violations' | 'major_violations' | 'non_compliant';
  principleScores: Map<string, number>; // Principle ID -> score (0-10)
  feedback: string;
  requiresRevision: boolean;
}

/**
 * Constitutional AI execution result
 */
export interface ConstitutionalAIResult {
  initialResponse: string;
  review: ConstitutionalReview;
  revisedResponse: string | null; // Null if no revision needed
  finalResponse: string;
  revisionCount: number;
  executionTime: number; // milliseconds
  metadata: {
    finalComplianceScore: number;
    violationCount: number;
    principleCount: number;
  };
}

/**
 * Constitutional AI configuration
 */
export interface ConstitutionalAIConfig {
  principles?: Principle[]; // Custom principles (default: healthcare)
  maxRevisions?: number; // Maximum revision attempts (default: 3)
  minComplianceScore?: number; // Minimum score to accept (default: 0.8)
  enableRevision?: boolean; // Enable automatic revision (default: true)
  llmProvider?: 'openai' | 'anthropic';
  model?: string;
  temperature?: number;
}

// ============================================================================
// HEALTHCARE CONSTITUTION (Default Principles)
// ============================================================================

/**
 * Healthcare-specific constitutional principles
 * 
 * Based on HIPAA, FDA, and medical best practices.
 */
export const HEALTHCARE_CONSTITUTION: Principle[] = [
  // Medical Accuracy (Weight: 3 = High Priority)
  {
    id: 'medical_accuracy',
    description:
      'Provide medically accurate information based on current clinical guidelines, peer-reviewed research, and evidence-based medicine. Never make definitive medical claims without sufficient evidence.',
    weight: 3,
    category: 'accuracy',
  },
  // No Diagnosis (Weight: 3 = High Priority)
  {
    id: 'no_diagnosis',
    description:
      'Never provide specific medical diagnoses without proper clinical examination. Always recommend professional consultation for diagnosis.',
    weight: 3,
    category: 'safety',
  },
  // Patient Safety (Weight: 3 = High Priority)
  {
    id: 'patient_safety',
    description:
      'Prioritize patient safety above all. When in doubt, recommend professional medical consultation. Never recommend dangerous interventions.',
    weight: 3,
    category: 'safety',
  },
  // Privacy & HIPAA (Weight: 2 = Medium-High Priority)
  {
    id: 'privacy_hipaa',
    description:
      'Respect patient privacy and handle sensitive health information appropriately. Never store or log PHI without proper authorization. Comply with HIPAA regulations.',
    weight: 2,
    category: 'privacy',
  },
  // Evidence-Based (Weight: 2 = Medium-High Priority)
  {
    id: 'evidence_based',
    description:
      'Base recommendations on evidence-based medicine. Cite sources when possible. Distinguish between established facts and emerging research.',
    weight: 2,
    category: 'accuracy',
  },
  // Cultural Sensitivity (Weight: 1 = Standard Priority)
  {
    id: 'cultural_sensitivity',
    description:
      'Be culturally sensitive and inclusive in healthcare recommendations. Consider diverse patient backgrounds and beliefs.',
    weight: 1,
    category: 'ethics',
  },
  // FDA Compliance (Weight: 2 = Medium-High Priority)
  {
    id: 'fda_compliance',
    description:
      'Comply with FDA regulations. Do not recommend off-label uses without clear disclaimers. Distinguish between approved and investigational treatments.',
    weight: 2,
    category: 'compliance',
  },
  // Transparency (Weight: 1 = Standard Priority)
  {
    id: 'transparency',
    description:
      'Be transparent about limitations, uncertainties, and the need for professional consultation. Clearly state when information is incomplete.',
    weight: 1,
    category: 'ethics',
  },
];

// ============================================================================
// DEFAULT CONFIGURATION
// ============================================================================

const DEFAULT_CONFIG: Required<
  Omit<ConstitutionalAIConfig, 'principles' | 'minComplianceScore'>
> & {
  minComplianceScore: number;
  principles: Principle[];
} = {
  principles: HEALTHCARE_CONSTITUTION,
  maxRevisions: 3,
  minComplianceScore: 0.8,
  enableRevision: true,
  llmProvider: 'openai',
  model: 'gpt-4-turbo-preview',
  temperature: 0.5,
};

// ============================================================================
// CONSTITUTIONAL AI CLASS
// ============================================================================

/**
 * Constitutional AI Agent
 * 
 * Implements self-critique and improvement based on constitutional principles.
 */
export class ConstitutionalAgent {
  private config: typeof DEFAULT_CONFIG;
  private logger;
  private tracing;
  private circuitBreaker;
  private llm: ChatOpenAI | ChatAnthropic;

  constructor(config: ConstitutionalAIConfig = {}) {
    this.config = {
      ...DEFAULT_CONFIG,
      ...config,
      principles: config.principles || HEALTHCARE_CONSTITUTION,
      minComplianceScore:
        config.minComplianceScore ?? DEFAULT_CONFIG.minComplianceScore,
    };

    this.logger = createLogger();
    this.tracing = getTracingService();
    this.circuitBreaker = getSupabaseCircuitBreaker();

    // Initialize LLM
    if (this.config.llmProvider === 'anthropic') {
      this.llm = new ChatAnthropic({
        modelName: this.config.model,
        temperature: this.config.temperature,
        anthropicApiKey: process.env.ANTHROPIC_API_KEY,
      });
    } else {
      this.llm = new ChatOpenAI({
        modelName: this.config.model,
        temperature: this.config.temperature,
        openAIApiKey: process.env.OPENAI_API_KEY,
      });
    }
  }

  /**
   * Generate response with constitutional review
   * 
   * @param query - User query
   * @param initialResponse - Initial response (optional, will generate if not provided)
   * @param context - Additional context for generation
   * @returns Complete constitutional AI result
   */
  async execute(
    query: string,
    initialResponse?: string,
    context: any[] = []
  ): Promise<ConstitutionalAIResult> {
    const executionStart = Date.now();
    const spanId = this.tracing.startSpan('ConstitutionalAgent.execute', undefined, {
      queryPreview: query.substring(0, 100),
      principleCount: this.config.principles.length,
    });

    try {
      // Step 1: Generate initial response (if not provided)
      let currentResponse =
        initialResponse || (await this.generateInitial(query, context));

      this.logger.debug('constitutional_ai_initial_generated', {
        responseLength: currentResponse.length,
        responsePreview: currentResponse.substring(0, 200),
      });

      // Step 2: Constitutional review
      let review = await this.constitutionalReview(currentResponse);
      let revisionCount = 0;

      // Step 3: Revise if needed (up to maxRevisions)
      while (
        review.requiresRevision &&
        revisionCount < this.config.maxRevisions &&
        this.config.enableRevision
      ) {
        this.logger.info('constitutional_ai_revising', {
          revisionCount: revisionCount + 1,
          complianceScore: review.complianceScore,
          violationCount: review.violations.length,
        });

        currentResponse = await this.reviseResponse(query, currentResponse, review);
        revisionCount++;

        // Re-review revised response
        review = await this.constitutionalReview(currentResponse);

        // If still below threshold after max revisions, accept best attempt
        if (
          review.complianceScore >= this.config.minComplianceScore ||
          revisionCount >= this.config.maxRevisions
        ) {
          break;
        }
      }

      const executionTime = Date.now() - executionStart;
      const finalResponse = currentResponse;

      const result: ConstitutionalAIResult = {
        initialResponse: initialResponse || currentResponse,
        review,
        revisedResponse:
          revisionCount > 0 ? currentResponse : null,
        finalResponse,
        revisionCount,
        executionTime,
        metadata: {
          finalComplianceScore: review.complianceScore,
          violationCount: review.violations.length,
          principleCount: this.config.principles.length,
        },
      };

      this.logger.infoWithMetrics('constitutional_ai_executed', executionTime, {
        revisionCount,
        finalComplianceScore: review.complianceScore,
        violationCount: review.violations.length,
        requiresRevision: review.requiresRevision,
      });

      this.tracing.endSpan(spanId, true);

      return result;
    } catch (error) {
      const executionTime = Date.now() - executionStart;
      this.logger.error(
        'constitutional_ai_execute_failed',
        error instanceof Error ? error : new Error(String(error)),
        {
          executionTime,
        }
      );

      this.tracing.endSpan(
        spanId,
        false,
        error instanceof Error ? error : new Error(String(error))
      );

      throw error;
    }
  }

  /**
   * Generate initial response
   * 
   * @param query - User query
   * @param context - Additional context
   * @returns Initial response
   */
  private async generateInitial(query: string, context: any[] = []): Promise<string> {
    const spanId = this.tracing.startSpan('ConstitutionalAgent.generateInitial', undefined, {
      queryLength: query.length,
      contextCount: context.length,
    });

    try {
      const contextText =
        context.length > 0
          ? `\n\nContext:\n${JSON.stringify(context, null, 2)}`
          : '';

      const prompt = `${query}${contextText}`;

      const response = await withRetry(
        async () => {
          return this.circuitBreaker.execute(async () => {
            return await this.llm.invoke([
              new SystemMessage(
                'You are a helpful healthcare assistant. Provide accurate, safe, and evidence-based information.'
              ),
              new HumanMessage(prompt),
            ]);
          });
        },
        {
          maxRetries: 2,
          initialDelayMs: 500,
          onRetry: (attempt, error) => {
            this.logger.warn('constitutional_ai_initial_retry', {
              attempt,
              error: error instanceof Error ? error.message : String(error),
            });
          },
        }
      );

      const initialResponse = response.content.toString();

      this.tracing.endSpan(spanId, true);

      return initialResponse;
    } catch (error) {
      this.logger.error(
        'constitutional_ai_initial_generation_failed',
        error instanceof Error ? error : new Error(String(error)),
        {}
      );

      this.tracing.endSpan(
        spanId,
        false,
        error instanceof Error ? error : new Error(String(error))
      );

      throw error;
    }
  }

  /**
   * Review response against constitutional principles
   * 
   * @param response - Response to review
   * @returns Constitutional review result
   */
  private async constitutionalReview(response: string): Promise<ConstitutionalReview> {
    const spanId = this.tracing.startSpan(
      'ConstitutionalAgent.constitutionalReview',
      undefined,
      {
        responseLength: response.length,
        principleCount: this.config.principles.length,
      }
    );

    const startTime = Date.now();

    try {
      const violations: Violation[] = [];
      const principleScores = new Map<string, number>();
      let totalWeightedScore = 0;
      let totalWeight = 0;

      // Review against each principle
      const reviewPromises = this.config.principles.map(async (principle) => {
        return this.reviewPrinciple(response, principle);
      });

      const principleReviews = await Promise.all(reviewPromises);

      for (const review of principleReviews) {
        const { principle, score, compliant, explanation } = review;

        principleScores.set(principle.id, score);
        totalWeightedScore += score * principle.weight;
        totalWeight += principle.weight * 10; // Max score is 10

        if (!compliant) {
          // Determine severity based on score and weight
          let severity: Violation['severity'] = 'low';
          if (score < 3) severity = 'critical';
          else if (score < 5) severity = 'high';
          else if (score < 7) severity = 'medium';

          violations.push({
            principleId: principle.id,
            principleDescription: principle.description,
            severity,
            explanation,
            weight: principle.weight,
          });
        }
      }

      // Calculate compliance score (0-1)
      const complianceScore =
        totalWeight > 0 ? totalWeightedScore / totalWeight : 0.5;

      // Determine overall compliance level
      let overallCompliance: ConstitutionalReview['overallCompliance'];
      const criticalViolations = violations.filter((v) => v.severity === 'critical').length;
      const highViolations = violations.filter((v) => v.severity === 'high').length;

      if (criticalViolations > 0 || complianceScore < 0.5) {
        overallCompliance = 'non_compliant';
      } else if (highViolations > 0 || complianceScore < 0.7) {
        overallCompliance = 'major_violations';
      } else if (violations.length > 0 || complianceScore < this.config.minComplianceScore) {
        overallCompliance = 'minor_violations';
      } else {
        overallCompliance = 'compliant';
      }

      const requiresRevision =
        overallCompliance !== 'compliant' &&
        complianceScore < this.config.minComplianceScore;

      const feedback = this.generateFeedback(violations, complianceScore);

      const duration = Date.now() - startTime;
      this.logger.infoWithMetrics('constitutional_ai_reviewed', duration, {
        complianceScore,
        violationCount: violations.length,
        criticalViolations,
        highViolations,
        overallCompliance,
      });

      this.tracing.endSpan(spanId, true);

      return {
        violations,
        complianceScore,
        overallCompliance,
        principleScores,
        feedback,
        requiresRevision,
      };
    } catch (error) {
      const duration = Date.now() - startTime;
      this.logger.error(
        'constitutional_ai_review_failed',
        error instanceof Error ? error : new Error(String(error)),
        { duration }
      );

      this.tracing.endSpan(
        spanId,
        false,
        error instanceof Error ? error : new Error(String(error))
      );

      // Graceful degradation: return lenient review
      return {
        violations: [],
        complianceScore: 0.8,
        overallCompliance: 'compliant',
        principleScores: new Map(),
        feedback: 'Review process encountered an error, assuming compliance',
        requiresRevision: false,
      };
    }
  }

  /**
   * Review response against a single principle
   * 
   * @param response - Response to review
   * @param principle - Principle to check
   * @returns Principle review result
   */
  private async reviewPrinciple(
    response: string,
    principle: Principle
  ): Promise<{
    principle: Principle;
    score: number;
    compliant: boolean;
    explanation: string;
  }> {
    const prompt = `
Evaluate this response against the principle:
"${principle.description}"

Response to evaluate:
${response}

Provide:
COMPLIANT: yes/no
SCORE: [0-10, where 10 is fully compliant]
EXPLANATION: [Brief explanation of compliance or violation]
    `.trim();

    try {
      const reviewResponse = await withRetry(
        async () => {
          return this.circuitBreaker.execute(async () => {
            return await this.llm.invoke([
              new SystemMessage(
                'You are a constitutional reviewer. Assess compliance objectively and rigorously.'
              ),
              new HumanMessage(prompt),
            ]);
          });
        },
        {
          maxRetries: 1,
          initialDelayMs: 300,
        }
      );

      const content = reviewResponse.content.toString();
      const compliant = content.toLowerCase().includes('compliant: yes');
      const scoreMatch = content.match(/SCORE:\s*(\d+)/i);
      const explanationMatch = content.match(/EXPLANATION:\s*(.+?)(?:\n|$)/is);

      const score = scoreMatch
        ? Math.min(10, Math.max(0, parseInt(scoreMatch[1])))
        : compliant
          ? 8
          : 5;
      const explanation = explanationMatch
        ? explanationMatch[1].trim()
        : compliant
          ? 'Response complies with principle'
          : 'Response may violate principle';

      return {
        principle,
        score,
        compliant,
        explanation,
      };
    } catch (error) {
      this.logger.warn('constitutional_ai_principle_review_failed', {
        principleId: principle.id,
        error: error instanceof Error ? error.message : String(error),
      });

      // Graceful degradation: assume compliance
      return {
        principle,
        score: 8,
        compliant: true,
        explanation: 'Review failed, assuming compliance',
      };
    }
  }

  /**
   * Revise response based on violations
   * 
   * @param query - Original query
   * @param response - Current response
   * @param review - Constitutional review
   * @returns Revised response
   */
  private async reviseResponse(
    query: string,
    response: string,
    review: ConstitutionalReview
  ): Promise<string> {
    const spanId = this.tracing.startSpan('ConstitutionalAgent.reviseResponse', undefined, {
      violationCount: review.violations.length,
      complianceScore: review.complianceScore,
    });

    try {
      const violationsText = review.violations
        .map(
          (v, i) =>
            `${i + 1}. ${v.principleDescription} (${v.severity}): ${v.explanation}`
        )
        .join('\n');

      const prompt = `
Original query: ${query}

Current response:
${response}

Constitutional violations found:
${violationsText}

Overall feedback: ${review.feedback}

Please revise the response to address these violations while:
1. Maintaining accuracy and helpfulness
2. Addressing all identified issues
3. Improving compliance with principles
4. Keeping the response natural and clear

REVISED RESPONSE:
      `.trim();

      const revised = await withRetry(
        async () => {
          return this.circuitBreaker.execute(async () => {
            return await this.llm.invoke([
              new SystemMessage(
                'You are an editor. Improve responses based on constitutional feedback while maintaining quality.'
              ),
              new HumanMessage(prompt),
            ]);
          });
        },
        {
          maxRetries: 2,
          initialDelayMs: 500,
          onRetry: (attempt, error) => {
            this.logger.warn('constitutional_ai_revision_retry', {
              attempt,
              error: error instanceof Error ? error.message : String(error),
            });
          },
        }
      );

      const revisedResponse = revised.content
        .toString()
        .replace(/REVISED RESPONSE:\s*/i, '')
        .trim();

      this.tracing.endSpan(spanId, true);

      return revisedResponse;
    } catch (error) {
      this.logger.error(
        'constitutional_ai_revision_failed',
        error instanceof Error ? error : new Error(String(error)),
        {}
      );

      this.tracing.endSpan(
        spanId,
        false,
        error instanceof Error ? error : new Error(String(error))
      );

      // Graceful degradation: return original response
      return response;
    }
  }

  /**
   * Generate feedback from violations
   * 
   * @param violations - Violations found
   * @param complianceScore - Overall compliance score
   * @returns Feedback string
   */
  private generateFeedback(violations: Violation[], complianceScore: number): string {
    if (violations.length === 0) {
      return 'Response is fully compliant with all constitutional principles.';
    }

    const criticalCount = violations.filter((v) => v.severity === 'critical').length;
    const highCount = violations.filter((v) => v.severity === 'high').length;

    if (criticalCount > 0) {
      return `Response has ${criticalCount} critical violation(s) requiring immediate revision.`;
    }

    if (highCount > 0) {
      return `Response has ${highCount} high-severity violation(s) that should be addressed.`;
    }

    return `Response has ${violations.length} minor violation(s). Compliance score: ${(
      complianceScore * 100
    ).toFixed(1)}%.`;
  }
}

File: backend/services/ai_engine/langgraph_compiler/patterns/constitutional_ai.py
"""
Constitutional AI Pattern

Implements Constitutional AI for self-critique and harmlessness alignment.

Paper: "Constitutional AI: Harmlessness from AI Feedback" (Bai et al., 2022)

Algorithm:
1. Generate initial response
2. Critique response against constitution (set of principles)
3. Revise response based on critique
4. Repeat if needed
5. Return aligned response

Constitution:
A set of principles/rules that guide the AI's behavior, such as:
- Be helpful, harmless, and honest
- Avoid harmful or unethical content
- Respect privacy and confidentiality
- Acknowledge uncertainty
- Follow medical/legal regulations

Use cases:
- Safety-critical applications (medical, legal)
- Content moderation
- Ethical AI alignment
- Regulatory compliance
"""

from typing import List, Dict, Any, Optional
from uuid import UUID
from dataclasses import dataclass
from enum import Enum
import openai

from ...graphrag.utils.logger import get_logger

logger = get_logger(__name__)


class CritiqueSeverity(str, Enum):
    """Severity of constitutional violation"""
    NONE = "none"           # No violations
    LOW = "low"             # Minor issues
    MEDIUM = "medium"       # Notable concerns
    HIGH = "high"           # Serious violations
    CRITICAL = "critical"   # Critical safety issues


@dataclass
class ConstitutionalPrinciple:
    """A single principle in the constitution"""
    id: str
    title: str
    description: str
    category: str  # e.g., "safety", "privacy", "accuracy", "ethics"
    is_required: bool = True  # If False, treated as guideline


@dataclass
class CritiqueResult:
    """Result of constitutional critique"""
    principle_id: str
    principle_title: str
    passes: bool
    severity: CritiqueSeverity
    issue_description: Optional[str] = None
    suggested_revision: Optional[str] = None


@dataclass
class ConstitutionalReview:
    """Complete constitutional review of a response"""
    original_response: str
    critiques: List[CritiqueResult]
    overall_passes: bool
    max_severity: CritiqueSeverity
    revised_response: Optional[str] = None
    revision_count: int = 0
    
    def get_failing_principles(self) -> List[CritiqueResult]:
        """Get all failing critiques"""
        return [c for c in self.critiques if not c.passes]
        
    def get_summary(self) -> str:
        """Get human-readable summary"""
        failing = self.get_failing_principles()
        if not failing:
            return "✅ All constitutional principles satisfied"
            
        issues = "\n".join([
            f"- [{c.severity.value}] {c.principle_title}: {c.issue_description}"
            for c in failing
        ])
        return f"⚠️ {len(failing)} constitutional issues found:\n{issues}"


class ConstitutionalAgent:
    """
    Implements Constitutional AI pattern for self-critique and alignment.
    
    This agent:
    1. Generates initial response
    2. Critiques against constitution
    3. Revises based on critique
    4. Iterates until aligned or max iterations
    """
    
    def __init__(
        self,
        agent_id: UUID,
        model: str = "gpt-4",
        constitution: List[ConstitutionalPrinciple] = None,
        critique_mode: str = "iterative",  # "iterative" or "single"
        max_revisions: int = 3,
        temperature: float = 0.3  # Lower temp for critique
    ):
        """
        Initialize Constitutional AI agent.
        
        Args:
            agent_id: Agent UUID
            model: LLM model to use
            constitution: List of principles to enforce
            critique_mode: "iterative" (revise multiple times) or "single" (one pass)
            max_revisions: Maximum revision iterations
            temperature: LLM temperature
        """
        self.agent_id = agent_id
        self.model = model
        self.constitution = constitution or self._default_constitution()
        self.critique_mode = critique_mode
        self.max_revisions = max_revisions
        self.temperature = temperature
        
        # Initialize OpenAI client
        from ...graphrag.config import get_embedding_config
        config = get_embedding_config()
        self.openai_client = openai.AsyncOpenAI(api_key=config.openai_api_key)
        
    async def critique(
        self,
        output: str,
        context: str = "",
        criteria: List[str] = None
    ) -> ConstitutionalReview:
        """
        Critique output against constitution.
        
        Args:
            output: Text to critique
            context: Context for evaluation
            criteria: Additional criteria beyond constitution
            
        Returns:
            ConstitutionalReview with critiques and revisions
        """
        logger.info(f"Constitutional critique started for agent {self.agent_id}")
        
        # Initialize review
        review = ConstitutionalReview(
            original_response=output,
            critiques=[],
            overall_passes=True,
            max_severity=CritiqueSeverity.NONE
        )
        
        current_output = output
        
        for iteration in range(self.max_revisions + 1):
            # Critique current output
            critiques = await self._critique_against_constitution(
                current_output,
                context,
                criteria
            )
            
            # Update review
            review.critiques = critiques
            review.overall_passes = all([c.passes for c in critiques])
            review.max_severity = self._get_max_severity(critiques)
            
            # If passes or single-pass mode, done
            if review.overall_passes or self.critique_mode == "single":
                if iteration > 0:
                    review.revised_response = current_output
                    review.revision_count = iteration
                break
                
            # Otherwise, revise
            if iteration < self.max_revisions:
                current_output = await self._revise_output(
                    current_output,
                    critiques,
                    context
                )
            else:
                # Max revisions reached
                review.revised_response = current_output
                review.revision_count = iteration
                logger.warning(
                    f"Constitutional alignment did not converge after {self.max_revisions} revisions"
                )
                
        logger.info(
            f"Constitutional critique completed: "
            f"passes={review.overall_passes}, severity={review.max_severity.value}, "
            f"revisions={review.revision_count}"
        )
        
        return review
        
    async def _critique_against_constitution(
        self,
        output: str,
        context: str,
        additional_criteria: List[str] = None
    ) -> List[CritiqueResult]:
        """Critique output against each constitutional principle"""
        critiques = []
        
        for principle in self.constitution:
            critique = await self._critique_principle(
                output,
                context,
                principle,
                additional_criteria
            )
            critiques.append(critique)
            
        return critiques
        
    async def _critique_principle(
        self,
        output: str,
        context: str,
        principle: ConstitutionalPrinciple,
        additional_criteria: List[str] = None
    ) -> CritiqueResult:
        """Critique output against a single principle"""
        criteria_text = ""
        if additional_criteria:
            criteria_text = "\nAdditional criteria:\n" + "\n".join(
                [f"- {c}" for c in additional_criteria]
            )
            
        prompt = f"""
Evaluate the following AI response against this constitutional principle:

Principle: {principle.title}
Description: {principle.description}
Category: {principle.category}
Required: {principle.is_required}
{criteria_text}

Context: {context}

AI Response:
{output}

Evaluate whether this response adheres to the principle. Provide:
1. "PASS" or "FAIL"
2. Severity: "none", "low", "medium", "high", or "critical"
3. Issue description (if any)
4. Suggested revision (if needed)

Format your response as JSON:
{{
  "passes": true/false,
  "severity": "none|low|medium|high|critical",
  "issue": "description of issue or null",
  "revision": "suggested revision or null"
}}
"""
        
        try:
            response = await self.openai_client.chat.completions.create(
                model=self.model,
                messages=[
                    {
                        "role": "system",
                        "content": "You are an expert at evaluating AI responses for constitutional alignment."
                    },
                    {"role": "user", "content": prompt}
                ],
                temperature=self.temperature,
                response_format={"type": "json_object"}
            )
            
            result_json = json.loads(response.choices[0].message.content)
            
            return CritiqueResult(
                principle_id=principle.id,
                principle_title=principle.title,
                passes=result_json.get("passes", True),
                severity=CritiqueSeverity(result_json.get("severity", "none")),
                issue_description=result_json.get("issue"),
                suggested_revision=result_json.get("revision")
            )
            
        except Exception as e:
            logger.error(f"Error critiquing principle {principle.id}: {e}")
            # Default to passing on error to avoid false blocks
            return CritiqueResult(
                principle_id=principle.id,
                principle_title=principle.title,
                passes=True,
                severity=CritiqueSeverity.NONE
            )
            
    async def _revise_output(
        self,
        output: str,
        critiques: List[CritiqueResult],
        context: str
    ) -> str:
        """Revise output based on critiques"""
        failing_critiques = [c for c in critiques if not c.passes]
        
        if not failing_critiques:
            return output
            
        # Build revision prompt
        issues_text = "\n".join([
            f"- {c.principle_title}: {c.issue_description}"
            for c in failing_critiques
        ])
        
        suggestions_text = "\n".join([
            f"- {c.principle_title}: {c.suggested_revision}"
            for c in failing_critiques
            if c.suggested_revision
        ])
        
        prompt = f"""
The following AI response has constitutional alignment issues:

Original Response:
{output}

Context: {context}

Issues identified:
{issues_text}

Suggested revisions:
{suggestions_text}

Please revise the response to address all issues while maintaining helpfulness and accuracy.

Provide ONLY the revised response, no explanations.
"""
        
        try:
            response = await self.openai_client.chat.completions.create(
                model=self.model,
                messages=[
                    {
                        "role": "system",
                        "content": "You are an expert at revising AI responses for constitutional alignment."
                    },
                    {"role": "user", "content": prompt}
                ],
                temperature=self.temperature
            )
            
            return response.choices[0].message.content.strip()
            
        except Exception as e:
            logger.error(f"Error revising output: {e}")
            return output  # Return original on error
            
    def _get_max_severity(self, critiques: List[CritiqueResult]) -> CritiqueSeverity:
        """Get maximum severity from critiques"""
        severity_order = [
            CritiqueSeverity.NONE,
            CritiqueSeverity.LOW,
            CritiqueSeverity.MEDIUM,
            CritiqueSeverity.HIGH,
            CritiqueSeverity.CRITICAL
        ]
        
        max_severity = CritiqueSeverity.NONE
        for critique in critiques:
            if severity_order.index(critique.severity) > severity_order.index(max_severity):
                max_severity = critique.severity
                
        return max_severity
        
    def _default_constitution(self) -> List[ConstitutionalPrinciple]:
        """Default constitution for medical AI"""
        return [
            ConstitutionalPrinciple(
                id="medical_accuracy",
                title="Medical Accuracy",
                description="Information must be medically accurate and evidence-based. Cite sources when possible.",
                category="accuracy",
                is_required=True
            ),
            ConstitutionalPrinciple(
                id="safety_first",
                title="Patient Safety",
                description="Never provide advice that could harm patients. Always recommend consulting healthcare professionals for medical decisions.",
                category="safety",
                is_required=True
            ),
            ConstitutionalPrinciple(
                id="privacy",
                title="Privacy & Confidentiality",
                description="Respect patient privacy. Never request or disclose personal health information inappropriately.",
                category="privacy",
                is_required=True
            ),
            ConstitutionalPrinciple(
                id="acknowledge_limits",
                title="Acknowledge Limitations",
                description="Acknowledge when information is uncertain or outside scope. Don't overstate confidence.",
                category="honesty",
                is_required=True
            ),
            ConstitutionalPrinciple(
                id="regulatory_compliance",
                title="Regulatory Compliance",
                description="Adhere to medical regulations and guidelines (FDA, EMA, etc.). Note off-label uses.",
                category="compliance",
                is_required=True
            ),
            ConstitutionalPrinciple(
                id="no_diagnosis",
                title="No Remote Diagnosis",
                description="Do not diagnose conditions remotely. Explain that diagnosis requires professional evaluation.",
                category="safety",
                is_required=True
            ),
            ConstitutionalPrinciple(
                id="cultural_sensitivity",
                title="Cultural Sensitivity",
                description="Be respectful of diverse cultural backgrounds and healthcare beliefs.",
                category="ethics",
                is_required=False  # Guideline
            ),
            ConstitutionalPrinciple(
                id="clarity",
                title="Clear Communication",
                description="Use clear, accessible language. Explain medical terms when used.",
                category="helpfulness",
                is_required=False  # Guideline
            )
        ]
        
    def add_principle(self, principle: ConstitutionalPrinciple) -> None:
        """Add a principle to the constitution"""
        self.constitution.append(principle)
        logger.info(f"Principle '{principle.title}' added to constitution")
        
    def remove_principle(self, principle_id: str) -> None:
        """Remove a principle from the constitution"""
        self.constitution = [p for p in self.constitution if p.id != principle_id]
        logger.info(f"Principle '{principle_id}' removed from constitution")

